Bugs 1164231, 1152697: Make long-press actions available as native accessibility actions

dusek opened the pull request:
[Bug 1164231 - Provide long-press actions as accessibility custom actions](https://bugzilla.mozilla.org/show_bug.cgi?id=1164231)
[Bug 1152697 - When saving a page to the reading list, post a VoiceOver announcement when the action finished successfully](https://bugzilla.mozilla.org/show_bug.cgi?id=1152697)

If you think the last commit is too wild, I can rework it to just
announce success on success and generic "Failed to add to Reading List"
otherwise. But I don't know of better solution right now except for
reworking `ReadingListStorageError` to represent all possible failures
as enum cases with suitable associated values, which is beyond the scope
of the proposed change.

Patches donated by [A11Y LTD.](http://a11y.ltd.uk)

dusek commented:
One more thing: the commit messages do contain some reasoning, if interested, you can read them too.

st3fan commented on Utils/Accessibility.swift:
```diff
@@ -0,0 +1,40 @@
+    func accessibilityCustomActionsForView(view: UIView) -> [UIAccessibilityCustomAction]?
+}
+
+public class Action {
```
I think this class needs a more descriptive name. How about `AccessibilityAction` instead?

dusek replied:
That is not the way I am thinking about it. `Action` really represents a generic action that has in principle nothing to do with accessibility, and from which both `UIAccessibilityCustomAction` and `UIAlertAction` can be generated.

dusek replied:
I know it is a bit unfortunate to put it in a file called `Accessibility.swift` then. But I could not think of a more suitable name :-)

thebnich replied:
How about `AccessibleAction` (note the subtle suggestion of Accessib**le** vs Accessib**ility**)? That is, rather than saying that the action is *for* accessibility, we're saying the action is *compatible with* accessibility. Would we ever want to create actions that aren't accessible?

The reason we're against simply using `Action` is that Swift is namespaced per-target, meaning `Action` becomes globally visible to all classes in that target (which is the majority of the project). `Action` might mean something else entirely in a different context (especially a non-UI one), so the class name should be as specific as possible to avoid ambiguity. At the very least, I'd change `Action` -> `UIAction` if `AccessibleAction` isn't suitable.

st3fan commented on ReadingList/ReadingListSQLStorage.swift:
```diff
@@ -94,6 +95,13 @@ class ReadingListSQLStorage: ReadingListStorage {
             }
         } else {
+            if let reason = statement.reason {
+                // TODO: this is ugly string matching, but checking for error SQLITE_CONSTRAINT does not say which constraint was violated, and even with checking for SQLITE_CONSTRAINT_UNIQUE extended result code, it is not possible to get which column's uniqueness was violated; so no idea at an "exact" way currently
```
The only unique field in the reading list database is the URL. So if you can check for `SQLITE_CONSTRAINT` then that is enough to know that the page was already in the reading list.

dusek replied:
I will see if I can dig up the last SQLite error. It was available as a private member of `Statement`, which I can't access, I will look if I can access it publicly in `Database`.

dusek replied:
@st3fan I can't seem to find a way to get at the numeric sqlite error code: `Database` has public *String* `lastError`, but no sqlite error code. `Statement` has private `status`, which is exactly what I would need, but it is private.

dusek commented:
Just a heads up that I will be finally addressing the review comments here in the coming days (hopefully before weekend).

st3fan commented:
@dusek i think we want to land this. are you happy with the state of these patches or do you want to improve them a bit more?

dusek commented:
@st3fan I just rebased and retested, seems to still work fine. If there is no hurry, I would like to look into improving them a little bit more w.r.t. review comments over the weekend.

dusek commented:
@st3fan Sorry for the delay. I finally got to addressing all review comments from you, @thebnich and @rnewman, including:
- making one delegate method return `Bool` instead of `Bool?` and documenting the return value
- getting rid of `URLBarAction` and related issues
- not using string matching to infer whether page could not be added to Reading List because it is already there

The only I review comment I did not accept was the one concerning the name of the class `Action` which I let be as it is, since it really is supposed to be an agnostic abstraction of different types of actions, hence it should be named generally too.

Please take a look and/or merge.

dusek commented:
One more thing:

> not using string matching to infer whether page could not be added to Reading List because it is already there

Here frankly I am waiting and hoping that someone else will provide a way to distinguish this, since that will be also needed when there will be visual notification implemented for adding page to Reading List ([Bug 1158503](https://bugzilla.mozilla.org/show_bug.cgi?id=1158503))

dusek commented on Client/Frontend/Browser/BrowserViewController.swift:
```diff
@@ -637,40 +665,41 @@ extension BrowserViewController: URLBarDelegate {
+               absoluteString = url.absoluteString,
+               result = profile.readingList?.createRecordWithURL(absoluteString, title: tab.title ?? "", addedBy: UIDevice.currentDevice().name)
+        {
+            switch(result) {
```
Btw. here I broke the syntax guidelines, but IMHO for good reason - including the beginning brace for `if` on the same line as `result = ...` would seem to place the `switch` at a lower indent level than the previous line. The way I did it, it is clear that the `switch` is inside the `if` since it is at a higher indent level than the preceding line (containing sole `{` on the same indent level as the `if`).

thebnich replied:
Nit: no parens with `switch` please

dusek replied:
done

thebnich commented on Client/Frontend/Browser/BrowserLocationView.swift:
```diff
@@ -131,7 +134,8 @@ class BrowserLocationView : UIView, ToolbarTextFieldDelegate {
         addSubview(readerModeButton)
         readerModeButton.isAccessibilityElement = true
-        readerModeButton.accessibilityLabel = NSLocalizedString("Reader Mode", comment: "Accessibility label for the reader mode button")
+        readerModeButton.accessibilityLabel = NSLocalizedString("Reader View", comment: "Accessibility label for the Reader View button")
```
Note that this change makes the reader tests fail, so please update them accordingly (I recommend re-running all tests whenever changing any accessibility labels since they're so tightly coupled to KIF).

dusek replied:
done

thebnich commented:
Wonderful, thanks!
